### DISCUSSION TITLE
Offline Standalone Mac Installer for Stable Diffusion Web UI (unofficial) 20230330

### DISCUSSION THREAD

**Discussion by brkirch** [brkirch/stable-diffusion-webui#1] [Announcements]
This is an offline standalone installer for a slightly modified version of AUTOMATIC1111's Stable Diffusion web UI (changes were applied to https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/a9fed7c364061ae6efb37f797b6b522cb3cf7aa2). All dependencies needed to start using Stable Diffusion (other than the model) are included and this does _not_ require Xcode command line tools to be installed.

Currently only Apple silicon Macs are supported. macOS 13.3 or later is highly recommended, older versions of macOS are untested.

<hr /><em>This discussion was created from the release <a href='https://github.com/brkirch/stable-diffusion-webui/releases/tag/20230330'>Offline Standalone Mac Installer for Stable Diffusion Web UI (unofficial) 20230330</a>.</em>

**Comment by jrittvo** [comment]
Is it a self-contained installation, relatively easy to remove, if necessary?  It will play nice with the Auto1111 installs we already have?  Does it install as a regular app, or more like at git repo setting up in whatever folder you run the installer from?

**Reply by brkirch** [reply]
It is completely self-contained. It installs in a folder (when the installer starts it lets you choose where to create the folder) and everything is put in that folder. It is still technically a git repo, but that is mostly due to how AUTOMATIC1111’s web UI was originally designed. The intention is to avoid needing `git pull` for upgrading, but I haven’t decided on how exactly upgrades will be handled yet. A launcher application is included in the folder in case you want to add it to the dock (moving the launcher outside of the installation folder is not currently supported, that may change in the future).

Since everything is put inside the installation folder, there is nothing to interfere with existing installations. Uninstalling is as simple as deleting the folder.

**Reply by MDMAchine** [reply]
Nice I'll check this out soon. Any modifications to the UI JSON? I know some things like certain resolutions we're not playing nice on Mac.

**Reply by jrittvo** [reply]
It is sounding like you intend this to be the "MAC Silicon" way to run AUTO1111, permanently - a parallel track just for us folk.  Is that how you see this, or is it more a temporary fork until things settle down?  A dedicated M1/M2/MX forkful would be a very tasty thing.  Might even encourage Apple to pull up a seat closer to the table and focus more on building out CoreML and MPS.  Regardless, thanks a bunch for this.

**Reply by brkirch** [reply]
@MDMAchine The main configuration changes are defaulting to frequent image generation preview updates (my testing shows that approx nn has virtually no performance impact, even when set to update on each step) and adding the upcast attention option next to model selection that should be checked when using a model based on SD 2.1.

As for resolutions, AFAIK the only selectable resolution that is broken is 2048x2048. With the Invoke AI cross attention optimization a lot of resolutions were broken, but the included webui-user.sh here defaults to sub-quadratic attention. Also there was a bug that created broken images for a lot of resolutions that was just recently fixed (macOS 13.3 fixed it, but there is a workaround for macOS 13.2.X that I’ve included also).

@jrittvo This is mostly intended to make for an easy way to install and launch AUTOMATIC1111’s web UI on Macs as well as hopefully providing a way to update but not break things as often. I do also plan to have important improvements and bug fixes here early, perhaps sometimes immediately after creating the PR for the main repo.

**Reply by MDMAchine** [reply]
Updated: Must of been the download, installer works.

**Comment by jrittvo** [comment]
I'm on an M1Pro, 16 GB, 13.2.1.  Before starting this new version, I copied over all the models, extensions, etc from the old install.  The new app started fine, seemed to pick up stuff from the pieces I added in, like grabbing ffmpeg to support text2video which we found over the weekend that Mac's can't support yet.  No unusual warnings or errors.  The web interface connected and at first glance looks fine.  It'll take me a bit now to put it through some paces.  I'll update after.  I have some run timings recorded from a few days ago.  If all is well. I'll rerun the same settings to get a comparison.  Hope you get some 13.3 testers soon.  That is where things were breaking down.

**Comment by jrittvo** [comment]
This install is idiot proof, so you succeeded at that goal.  The old install itself wasn't bad, but the prerequisites were hard on a novice and installing xCode, command line tools, takes up a ton of space, and learning git, pip, conda, whatever was all new to me and not anything a normal person would want to touch.  This new one was like using an app store app.  Maybe just needs a warning that it'll seem like it isn't doing anything for a minute or two while it sets itself up.

**Reply by MDMAchine** [reply]
I agree! This will be most helpful to those who cant, or doesn't have the time to config a setup.

I will note that i tried to install a extension from within and it failed (image browser in this case). 
```
GitCommandError: Cmd('git') failed due to: exit code(128) cmdline: git clone -v -- https://github.com/AlUlkesh/stable-diffusion-webui-images-browser.git /Volumes/External Drive/Stable Diffusion/AUTOMATIC1111/stable-diffusion-webui/tmp/stable-diffusion-webui-images-browser stderr: 'Cloning into '/Volumes/External Drive/Stable Diffusion/AUTOMATIC1111/stable-diffusion-webui/tmp/stable-diffusion-webui-images-browser'... warning: templates not found in /opt/homebrew/Cellar/git/2.40.0/share/git-core/templates git: 'remote-https' is not a git command. See 'git --help'. '
[object Object]
```

Also for the heck of it tried to run ngrok and that does not work.

**Reply by jrittvo** [reply]
A few extensions were updated the past few days for whatever the Gradio changes were.  This "app" is before the Gradio changes I think.  That's going to be a problem for a while, unfortunately.

**Comment by jrittvo** [comment]
Had a crash of a browser tab while going through all the settings, on like the last part.  Needed to go through them all again, this time saving along the way.  Crashed the tab at the same place a few times, but eventually made it through.  Feels like a Gradio issue.  Browser is Firefox.  Updated my extensions.  Gonna fully restart the app next and then try some generating.

**Comment by jrittvo** [comment]
Seems to be all ok here.  It starts up noticeably faster than it did before, or maybe I don't have much else running and it was able to cache more of the app in memory.  It's generating fine with Hi Res fix om, using a 1.5 fp16 model and using Euler and DPM++ SDE Karras, which are the 2 I stick to.  I'll check speed next.  Then I'm off for a nap.  This was an up all night, but well worth it.  I really hope this helps the folks using 13.3.  There wasn't much method to the madness.  So many different versions of packages.  I use some complicated audio apps in Debian, and one programming group does not release through the normal repo channels.  All their apps are what I think is called "statically linked".  All the dependencies are included with each release of the apps.  Whatever else happens with general system updates doesn't touch the apps.  It is a rational way to handle a fast changing world.

**Comment by kit-hub** [comment]
When I try to open the installer [install-web-ui.7z] it says 

"Install Web UI” is damaged and can’t be opened. You should move it to the Bin."

I'm on 13.3 and have tried redownloading, different unarchivers. Looks like @MDMAchine had a similar issue - anything else I can try?

**Reply by jrittvo** [reply]
Did you unzip the installer file first?  it extracts to a .app

**Reply by kit-hub** [reply]
I was using the Keka application but it seems to have worked using p7zip - thanks for the encouragement!

**Reply by jrittvo** [reply]
I'm getting text to image generation speeds about 10% to 20% faster than before, on a straight 512x512 with no hi res or anything else fancy.  That is the only benchmark I made before.  It is about the same speed I get running diffusers 0.14.0 and  torch 2.1.0 dev straight in python 3.10.  For me, with just python, the change from pytorch 1.13.1 to 2.0.0 also had gotten me a 20% bump in speed.  Mochi Diffusion using Swift and CoreMl is still the fastest, though.

**Reply by jrittvo** [reply]
Strange because Keka usually works when other ones don't.  Glad you get it going.  Now you're onto the real test.

**Reply by brkirch** [reply]
That issue is fixed now, redownload if you were having this problem and couldn't resolve it.

**Comment by jrittvo** [comment]
I was just able to pull in an extension from the extensions tab (off github) ok, but I don't know which place it ran git from.  MD, your git was running from homebrew.  There is a git in the new package.  Maybe I can see in the console which mine used.  I'm curious why it went to homebrew for you.

**Comment by MDMAchine** [comment]
So far so good, added --listen so i can use another computer to make the generations. Outside of my situation with the extensions not pulling, and ngrok not working even if I add pip install xxxx everything so far is working as intended.

Now an issue I saw as well with my optimized mac version was certain processes "worked" but resulted in a radical degrade in performance afterwards. One that I knew to test was upscaling with default parameters using the SwinIR 4x upscaler. It results in a corrupted image, and any procedure afterwards is painfully slow (I ran a image generation that normally would take 10 seconds and its currently at 3:19).

Another instance that I havent tested but will soon, is if I ran hi-res fix with anything other than latent the same thing would happen. Granted I have not received an update for 13.3 yet and I didnt want to do the dev route.

**Reply by jrittvo** [reply]
I’m doing some runs based on what you say above with Hi Res fix and different upscalers. Right off the bat, SwinIR4x is slow for sure, but I’m not getting any residual effect on the following run. But we may be using very different settings. I’ll spell mine out after I try one more variable.

**Reply by MDMAchine** [reply]
Its strange, whereas under normal run the gpu will have a load for the entire generation process. After this happens it'll spike, drop for 10 seconds, spike and drop, etc...

This would happen sometimes rendering higher resolution coreml stuff too. Never could pin it down.

**Comment by jrittvo** [comment]
I'm running each batch of settings twice in a row and using the times just from the second run.  Using a batch count of 2 and dividing the batch time in half.

My baseline is a 512x512 with 24 steps using Euler.  Takes 22 seconds.  When I just change the size for the straight run to 768x768, the time goes up to 75 seconds.  So that feels like a geometric change, and it's big.

Going back to an initial 512x512 and setting the Hi Res Fix at 1.5x to get to the 768x768 again, with 14 steps for the Hi Res pass, using "None" as the upscaler, I hit 65 seconds.  Again. the time seems to be scaling almost geometrically with the size increase.  Switching to Latent Bicubic Anti-aliased is the same 65 seconds.  With ESRGAN4x, which is what I found to be the best compromise between time and quality when I dug into this last week, the time goes up to 72 seconds, a 10% bump.  With SwinIR4x, the time jumps to 102 seconds.  I see no improvement at all, so on my rig it is just burning time.  It doesn't screw up anything afterwards, but I have 16GB and I think you said you have 8, and maybe it isn't flushing memory well but I have more to hide the issue.  FWIW, I remember R-ESRGAN4x being terribly slow and no better quality, just like SwinR4x.  I think I found R-ESRGan2x to be the same as ESRGAN2x in speed, but a bit sharper in a way that seemed less natural to me.  It might be better with larger images that can handle more sharpening.

Then I went back to the ESRGAN4x and changed the number of Hi Res steps, with everything else kept the same.  Dropping from 14 HR steps to 4, the time dropped to 44 seconds, which makes sense.  The iterations per second stay more or less constant and the time varies directly with the number of steps.  But the time saved with just 4 HR steps left a somewhat fuzzy image with some noise artifacts.  Going the other way, up to 24 HR steps (the same as the number of first pass steps) the time went up to 100 seconds.  Again linearly proportional to the increase in steps, and with no improvement in image quality that I could see.  In fact, it starts to change faces and other details from the first pass.  So if you tell it to also save before doing the Hi Res fix, you get 2 minor variations for the price of 1, but at the 2 different resolutions.  Seems better to run the batch again at the middle Hi Res step count and get 2 images from different seeds.

Now this is all with the Hi Res Fix.  The upscaling pipeline might be completely different.  I never tried it and don't plan to.  Staying with the Hi Res Fix keeps all the settings closer together for me and I find that less stressful to deal with.  Frankly, there are way too many options for my temperament.  I've been removing things from the interface that I don't use once I more or less decide on what I do use.

I've also noticed the GPU (and memory) spiking and falling, but within individual runs, like they are breathing, even when it seems they must be doing a tight loop over and over.  Doesn't happen at all with the CPU.  Doesn't make sense, and I don't know if I trust the app I use to monitor that stuff to be accurate.  The hiccuping while in a run is actually worst with Mochi for me, and the overall levels spike highest there.  Watching something ht 97% scares me.

**Reply by MDMAchine** [reply]
Yeah Im thinking ram and memory management is the likely culprit. In past tests it happens even if using lowvram or medvram parameters as well,

Using hi-res fix is a much more tedious process with just 8gb apparently. If I try to use it and go above 2.0 I get:
`RuntimeError: MPS backend out of memory (MPS allocated: 4.16 GB, other allocations: 4.02 GB, max allowed: 9.07 GB). Tried to allocate 1012.50 MB on private pool. Use PYTORCH_MPS_HIGH_WATERMARK_RATIO=0.0 to disable upper limit for memory allocations (may cause system failure).
`
It might run once successfully but thats it. So I have to use the upscaling and even then I threw in BSRGAN because its to most efficient. I can always use another upscaler (or CoreML upscalers) later on.

**Comment by jrittvo** [comment]
Have you played with "tiling"?  I'm not really sure what it does, but taking the word literally, it may work on sections and then stitch them together.  Might still be slow, but there could be a tile size that was rock stable given a particular amount of memory.  If it's available  on the tab that does batching, you could run it over night when time doesn't matter as much.

**Reply by MDMAchine** [reply]
I believe tiling is just the image can be tiles seamlessly.
[Like this](https://i0.wp.com/stable-diffusion-art.com/wp-content/uploads/2023/03/image-59.png?resize=768%2C768&ssl=1)

**Comment by jrittvo** [comment]
Yep.  Now that you mention it, I remember reading that 2 or 3 days ago in regard to a pattern making model I was converting.

**Comment by narimangafurov** [comment]
thank you!
